pkg/kademlia/testrouting/testrouting.go

```diff
@@ -0,0 +1,287 @@
+	return &Table{
+		self:            self,
+		bucketSize:      bucketSize,
+		cacheSize:       cacheSize,
+		allowedFailures: allowedFailures,
+		nodes:           map[storj.NodeID]*nodeData{},
+	}
+}
+
+// make sure the Table implements the right interface
+var _ dht.RoutingTable = (*Table)(nil)
+
+// K returns the Table's routing depth, or Kademlia k value
+func (t *Table) K() int { return t.bucketSize }
+
+// CacheSize returns the size of
+func (t *Table) CacheSize() int { return t.cacheSize }
+
+// ConnectionSuccess should be called whenever a node is successfully connected
+// to. It will add or update the node's entry in the routing table.
+func (t *Table) ConnectionSuccess(node *pb.Node) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// don't add ourselves
+	if node.Id == t.self {
+		return nil
+	}
+
+	// if the node is already here, update everything but its placement order
+	if cell, exists := t.nodes[node.Id]; exists {
+		cell.node = node
+		cell.timestamp = time.Now()
+		cell.fails = 0
+		return nil
+	}
+
+	// add unconditionally (it might be going into a replacement cache)
+	t.nodes[node.Id] = &nodeData{
+		node:      node,
+		ordering:  t.counter,
+		timestamp: time.Now(),
+		fails:     0,
+	}
+	t.counter += 1
+
+	// prune replacement caches
+	t.makeTree().walkLeaves(func(b *bucket) {
+		if len(b.cache) <= t.cacheSize {
+			return
+		}
+		for _, node := range b.cache[:len(b.cache)-t.cacheSize] {
+			delete(t.nodes, node.node.Id)
+		}
+	})
+	return nil
+}
+
+// ConnectionFailed should be called whenever a node can't be contacted.
+// If a node fails more than allowedFailures times, it will be removed from
+// the routing table. The failure count is reset every successful connection.
+func (t *Table) ConnectionFailed(node *pb.Node) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// if the node exists and the failure is with the address we have, record
+	// a failure
+	if data, exists := t.nodes[node.Id]; exists &&
+		addressEqual(data.node.Address, node.Address) {
+		data.fails += 1
+
+		// if we've failed too many times, remove the node
+		if t.allowedFailures < data.fails {
+			delete(t.nodes, node.Id)
+		}
+	}
+	return nil
+}
+
+// FindNear will return up to limit nodes in the routing table ordered by
+// kademlia xor distance from the given id.
+func (t *Table) FindNear(id storj.NodeID, limit int) ([]*pb.Node, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// find all non-cache nodes
+	nodes := make([]*nodeData, 0, len(t.nodes))
+	t.makeTree().walkLeaves(func(b *bucket) {
+		nodes = append(nodes, b.nodes...)
+	})
+
+	// sort by distance
+	sort.Sort(nodeDataDistanceSorter{self: id, nodes: nodes})
+
+	// return up to limit nodes
+	if len(nodes) < limit {
+		limit = len(nodes)
+	}
+	rv := make([]*pb.Node, 0, limit)
+	for _, data := range nodes[:limit] {
+		rv = append(rv, data.node)
+	}
+	return rv, nil
+}
+
+// Local implements the dht.RoutingTable interface
+func (t *Table) Local() pb.Node {
```

Comment:
In the interface definition we called `Local` what you are calling `Self` below. We also had it storing the Node and not just the ID.

Reply:
right, so, this routing table doesn't have the node, because the routing table is actually going to be stale about self's address. i was saving this for another pr but i think we should change the interface of every routing table to not keep track of the self node, and only the self id.

Reply:
well, in the long term it will be the nat traversal system that will be responsible for keeping track of self's address. in the short term it's just the config system that knows the public address. planning for the public address changing is what i have in mind

Reply:
well self is used for more than just address, we are using it right now for updating restrictions to be broadcast. Is there somewhere else we should keep track of metadata about a node for the network?

Reply:
i'm leaning towards yes. the main issue is that once we get to having automatic nat traversal, a node won't necessarily know what its public address even is. we should store metadata and node id and stuff somewhere, but having to store the address and potentially returning it in kademlia results has started to worry me

Reply:
how do you feel about storing self in the routing table struct, but not in the routing table itself? the routing table interface could support updating and retrieving the self node, but it wouldn't get returned in node results or queries?

Reply:
Yea that's fine with me. Maybe Routing Table is an all together bad place? maybe just on the top level Kademlia struct ? We just need a `Self` somewhere to track data about a node.